File: crates/nomoreide-core/src/vultr_provider.rs
//! Vultr as host provider #1.
//!
//! The Rust half of `src/core/vultr-provider.ts`. Everything vendor-specific
//! about a Vultr machine stops here: the caller above sees only the neutral
//! `HostInstance`.

use crate::providers::api_base::provider_api_host;
use crate::providers::host::{HostInstance, HostInstanceSpecs, HostInstanceState};
use crate::vultr_manager::api_base;
use serde_json::Value;

pub const VULTR_ACTIONS: &[&str] = &["start", "halt", "reboot"];

/// The two actions that stop a machine that may be serving something. The
/// dashboard asks before running one; `start` needs no such question.
pub const PRODUCTION_AFFECTING_ACTIONS: &[&str] = &["halt", "reboot"];

/// Vultr builds an instance with either `root` enabled or a sudo-capable
/// `linuxuser` instead, and says which through `user_scheme`. A target that
/// names the wrong one is a login that always fails.
const ROOT_USER: &str = "root";
const LIMITED_USER: &str = "linuxuser";
const LIMITED_SCHEME: &str = "limited";

/// The two spellings of "not assigned yet". Vultr reports an address it has not
/// given a machine as one of these rather than as nothing, for either family,
/// so neither is an address.
const UNASSIGNED: [&str; 2] = ["0.0.0.0", "::"];

/// The one vendor word that decides a machine's state.
///
/// Vultr spreads the answer over three fields, and at any moment exactly one of
/// them is the deciding one. The subscription's own lifecycle outranks
/// everything — a suspended machine still reports `power_status: "running"`,
/// and showing that as running would be a lie. Once the subscription is
/// `active`, a machine mid-install or locked is not yet usable whatever its
/// power says. Power is the last word rather than the first.
///
/// Only `installingbooting` and `locked` are deciding server states; anything
/// else there, `none` included, falls through to power.
fn deciding_word(raw: &Value) -> String {
    let status = text(raw, "status");
    if status != "active" {
        return status;
    }
    match text(raw, "server_status").as_str() {
        // The vendor's own word is `installingbooting`; the one reported is the
        // half a person reads.
        "installingbooting" => "installing".to_string(),
        "locked" => "locked".to_string(),
        _ => text(raw, "power_status"),
    }
}

/// That word onto the five neutral states.
///
/// An unrecognised word is `Unknown` rather than a guess: the alternative is a
/// dashboard that confidently says "running" about a machine nobody classified.
/// `raw_state` carries the word itself, so the UI can say what it did not know.
pub fn state_of(word: &str) -> HostInstanceState {
    match word {
        "running" => HostInstanceState::Running,
        "stopped" => HostInstanceState::Stopped,
        "installing" | "locked" | "pending" | "resizing" => HostInstanceState::Provisioning,
        "suspended" => HostInstanceState::Error,
        _ => HostInstanceState::Unknown,
    }
}

fn text(raw: &Value, key: &str) -> String {
    raw.get(key)
        .and_then(Value::as_str)
        .unwrap_or_default()
        .to_string()
}

/// A string field, absent when the vendor left it empty.
///
/// The vendor blanks `region`, `plan` and `os` on a machine it has not finished
/// placing, and an empty string in the answer would be reported by the
/// dashboard as a region whose name is nothing.
fn optional_text(raw: &Value, key: &str) -> Option<String> {
    let value = text(raw, key);
    (!value.is_empty()).then_some(value)
}

/// A size, absent when the vendor reports none — which it spells as a zero.
fn optional_number(raw: &Value, key: &str) -> Option<u64> {
    raw.get(key)
        .and_then(Value::as_u64)
        .filter(|value| *value != 0)
}

/// One address, or nothing.
///
/// Shared by both families deliberately: the sentinels are the vendor's, not
/// IPv4's, and a v6 field carrying `::` means exactly what a v4 field carrying
/// `0.0.0.0` means.
fn address(raw: &Value, key: &str) -> Option<String> {
    let value = text(raw, key);
    let trimmed = value.trim();
    (!trimmed.is_empty() && !UNASSIGNED.contains(&trimmed)).then(|| trimmed.to_string())
}

/// The vendor's ISO timestamp as epoch milliseconds.
fn epoch_ms(value: &str) -> Option<i64> {
    chrono::DateTime::parse_from_rfc3339(value)
        .ok()
        .map(|parsed| parsed.timestamp_millis())
}

pub fn instance_from_raw(raw: &Value) -> HostInstance {
    let word = deciding_word(raw);
    let hostname = optional_text(raw, "hostname");
    let id = text(raw, "id");

    HostInstance {
        // A machine with no label is shown by the name it answers to, and one
        // with neither by its id — an empty row in the list is worse than a
        // technical one, and there is always an id.
        label: optional_text(raw, "label")
            .or_else(|| hostname.clone())
            .unwrap_or_else(|| id.clone()),
        id,
        state: state_of(&word),
        raw_state: word.clone(),
        ipv4: address(raw, "main_ip"),
        ipv6: address(raw, "v6_main_ip"),
        hostname,
        region: optional_text(raw, "region"),
        plan: optional_text(raw, "plan"),
        os: optional_text(raw, "os"),
        specs: HostInstanceSpecs {
            vcpus: optional_number(raw, "vcpu_count"),
            memory_mb: optional_number(raw, "ram"),
            disk_gb: optional_number(raw, "disk"),
        },
        tags: raw
            .get("tags")
            .and_then(Value::as_array)
            .map(|tags| {
                tags.iter()
                    .filter_map(Value::as_str)
                    .map(str::to_string)
                    .collect()
            })
            .unwrap_or_default(),
        created_at: epoch_ms(&text(raw, "date_created")),
        default_user: if text(raw, "user_scheme") == LIMITED_SCHEME {
            LIMITED_USER.into()
        } else {
            ROOT_USER.into()
        },
    }
}

/// The manifest the dashboard renders a tab from.
pub fn manifest() -> Value {
    serde_json::json!({
        "id": crate::vultr_auth::VULTR_PROVIDER_ID,
        "name": "Vultr",
        "kind": "host",
        "strings": {
            "en": {
                "action.start": "Start",
                "action.start.done": "Instance starting.",
                "action.halt": "Stop",
                "action.halt.done": "Instance stopping.",
                "action.halt.confirmTitle": "Stop this instance?",
                "action.halt.confirm": "The machine powers off immediately. Anything running on it stops.",
                "action.reboot": "Reboot",
                "action.reboot.done": "Instance rebooting.",
                "action.reboot.confirmTitle": "Reboot this instance?",
                "action.reboot.confirm": "The machine restarts immediately. Anything running on it stops."
            },
            "zh": {
                "action.start": "启动",
                "action.start.done": "实例正在启动。",
                "action.halt": "停止",
                "action.halt.done": "实例正在停止。",
                "action.halt.confirmTitle": "停止该实例？",
                "action.halt.confirm": "机器将立即关机，其上运行的一切都会停止。",
                "action.reboot": "重启",
                "action.reboot.done": "实例正在重启。",
                "action.reboot.confirmTitle": "重启该实例？",
                "action.reboot.confirm": "机器将立即重启，其上运行的一切都会停止。"
            }
        },
        "actions": VULTR_ACTIONS,
        "productionAffecting": PRODUCTION_AFFECTING_ACTIONS,
        // The API only. An adopted instance is reached over SSH by its own
        // transport, which never goes through this fetch — so no instance
        // address belongs here.
        "api": { "hosts": [provider_api_host(&api_base())] },
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn word(status: &str, power: &str, server: &str) -> String {
        deciding_word(&json!({
            "status": status, "power_status": power, "server_status": server
        }))
    }

    #[test]
    fn the_subscription_outranks_the_power_field() {
        assert_eq!(word("suspended", "running", "ok"), "suspended");
        assert_eq!(state_of("suspended"), HostInstanceState::Error);
        assert_eq!(word("pending", "stopped", "none"), "pending");
    }

    #[test]
    fn a_machine_still_installing_or_locked_is_not_running() {
        assert_eq!(word("active", "running", "installingbooting"), "installing");
        assert_eq!(word("active", "running", "locked"), "locked");
        assert_eq!(state_of("installing"), HostInstanceState::Provisioning);
    }

    /// Everything else a server status can say falls through to power — `none`
    /// included, which is what a machine that finished installing reports.
    #[test]
    fn any_other_server_state_defers_to_power() {
        assert_eq!(word("active", "running", "none"), "running");
        assert_eq!(word("active", "running", "something_new"), "running");
        assert_eq!(word("active", "stopped", "ok"), "stopped");
    }

    #[test]
    fn an_unmapped_word_is_unknown_rather_than_a_guess() {
        assert_eq!(word("something_new", "running", "ok"), "something_new");
        assert_eq!(state_of("something_new"), HostInstanceState::Unknown);
        assert_eq!(word("active", "something_new", "ok"), "something_new");
    }

    #[test]
    fn an_unassigned_address_is_not_an_address() {
        let raw = json!({ "id": "i", "main_ip": "0.0.0.0", "hostname": "h" });
        assert_eq!(instance_from_raw(&raw).ipv4, None);
    }

    #[test]
    fn a_blank_label_falls_back_to_the_hostname() {
        let raw = json!({ "id": "i", "label": "", "hostname": "box.example.test" });
        assert_eq!(instance_from_raw(&raw).label, "box.example.test");
    }
}

File: src/core/vultr-provider.ts
import type {
  HostInstance,
  HostInstanceState,
  HostProvider,
  HostProviderActions,
  HostProviderManifest,
  HostSshTarget,
} from "./providers/host-provider.js";
import {
  PRODUCTION_AFFECTING_ACTIONS,
  VULTR_ACTIONS,
  type VultrActions,
} from "./vultr-actions.js";
import { VULTR_PROVIDER_ID } from "./vultr-auth.js";
import { providerApiHost } from "./providers/api-base.js";
import { vultrApiBase, type VultrInstance, type VultrManager } from "./vultr-manager.js";

/**
 * Vultr as `HostProvider` implementation #1 — the provider that tests whether
 * the *second* contract was shaped by the problem or by the first one.
 *
 * A pure adapter, like `vercel-provider.ts` and `cloudflare-provider.ts`: every
 * method delegates to `VultrManager` or `VultrActions` and only reshapes the
 * result.
 *
 * What writing it surfaced:
 *
 * 1. **A host provider is not repo-scoped.** `DeployProvider` needs a working
 *    directory to answer anything, because a deployment belongs to the project
 *    a repository is linked to. An instance belongs to an account. So there is
 *    no project-resolution ladder here, no link file, no `repoUrl` hook — the
 *    whole of `providers/project-resolution.ts` is simply not part of this
 *    contract, which is why the split into two contracts was right.
 * 2. **`toSshTarget` needs the machine's login, not just its address.** Hence
 *    `HostInstance.defaultUser`: Vultr builds an instance with either `root` or
 *    an unprivileged `linuxuser`, and a target that omits it is one `ssh` tries
 *    as the *local* username, which fails for every Vultr instance.
 * 3. **The shared credential layer carried over untouched.** Vultr reuses
 *    `providers/credentials.ts` exactly, despite being neither a deploy
 *    platform nor an OAuth one — the first evidence that layer generalized
 *    beyond the contract it was extracted from.
 */

export const VULTR_MANIFEST: HostProviderManifest = {
  id: VULTR_PROVIDER_ID,
  name: "Vultr",
  kind: "host",
  // Three action names that share not one word with either deploy provider,
  // which is the clearest evidence for why these strings belong to the plugin:
  // no host catalogue written for Vercel could ever have contained "halt".
  // No `scope.label` — a host provider has no scopes.
  strings: {
    en: {
      "action.start": "Start",
      "action.start.done": "Instance starting.",
      "action.halt": "Stop",
      "action.halt.done": "Instance stopping.",
      "action.halt.confirmTitle": "Stop this instance?",
      "action.halt.confirm":
        "The machine powers off immediately. Anything running on it stops.",
      "action.reboot": "Reboot",
      "action.reboot.done": "Instance rebooting.",
      "action.reboot.confirmTitle": "Reboot this instance?",
      "action.reboot.confirm":
        "The machine restarts immediately. Anything running on it stops.",
    },
    zh: {
      "action.start": "启动",
      "action.start.done": "实例正在启动。",
      "action.halt": "停止",
      "action.halt.done": "实例正在停止。",
      "action.halt.confirmTitle": "停止该实例？",
      "action.halt.confirm": "机器将立即关机，其上运行的一切都会停止。",
      "action.reboot": "重启",
      "action.reboot.done": "实例正在重启。",
      "action.reboot.confirmTitle": "重启该实例？",
      "action.reboot.confirm": "机器将立即重启，其上运行的一切都会停止。",
    },
  },
  actions: [...VULTR_ACTIONS],
  productionAffecting: [...PRODUCTION_AFFECTING_ACTIONS],
  // The API only. An adopted instance is reached over SSH by `ssh-servers.ts`,
  // which is the host's own transport and never goes through this fetch — so no
  // instance address belongs here.
  api: { hosts: [providerApiHost(vultrApiBase())] },
};

/**
 * Vultr's three-field state onto the five neutral ones.
 *
 * Vultr describes a machine with `status` (the subscription's lifecycle),
 * `power_status` (whether the VM is switched on) and `server_status` (how far
 * the OS install has got). Only their combination says whether the machine can
 * be reached: an `active` instance that is `running` is still unreachable while
 * `server_status` is `installingbooting`.
 *
 * `suspended` has no neutral equivalent — it is Vultr locking the machine, and
 * the user has to go and deal with it — so it maps to `error` and survives
 * verbatim in `rawState`, the same pressure valve `skipped` used on Cloudflare.
 */
export function toInstanceState(instance: VultrInstance): HostInstanceState {
  switch (instance.status) {
    case "pending":
    case "resizing":
      return "provisioning";
    case "suspended":
      return "error";
    case "active":
      break;
    default:
      return "unknown";
  }
  if (instance.serverStatus === "installingbooting") return "provisioning";
  if (instance.serverStatus === "locked") return "provisioning";
  if (instance.powerStatus === "running") return "running";
  if (instance.powerStatus === "stopped") return "stopped";
  return "unknown";
}

/**
 * The word Vultr's own dashboard shows, so `rawState` reads as something the
 * user recognizes rather than as a field dump: the lifecycle status while the
 * instance is not yet `active`, then the install stage, then the power state.
 */
export function toRawState(instance: VultrInstance): string {
  if (instance.status !== "active") return instance.status;
  if (instance.serverStatus === "installingbooting") return "installing";
  if (instance.serverStatus === "locked") return "locked";
  return instance.powerStatus || "unknown";
}

/**
 * Vultr's `user_scheme` onto the account name actually present on the machine.
 * `limited` provisions a sudo-capable `linuxuser` instead of enabling `root`.
 */
export function toDefaultUser(instance: VultrInstance): string {
  return instance.userScheme === "limited" ? "linuxuser" : "root";
}

export function toHostInstance(instance: VultrInstance): HostInstance {
  return {
    id: instance.id,
    // A Vultr instance may be created with no label at all, in which case its
    // hostname is the only human-readable thing about it.
    label: instance.label || instance.hostname || instance.id,
    state: toInstanceState(instance),
    rawState: toRawState(instance),
    ipv4: instance.mainIp,
    ipv6: instance.v6MainIp,
    hostname: instance.hostname || undefined,
    region: instance.region || undefined,
    plan: instance.plan || undefined,
    os: instance.os || undefined,
    specs: {
      vcpus: instance.vcpuCount || undefined,
      memoryMb: instance.ramMb || undefined,
      diskGb: instance.diskGb || undefined,
    },
    tags: instance.tags,
    createdAt: instance.createdAt,
    defaultUser: toDefaultUser(instance),
  };
}

/**
 * The instance as an SSH host — the payoff this contract exists for.
 *
 * Exported and pure so the mapping is testable on its own, and shared by the
 * provider below.
 */
export function toVultrSshTarget(instance: HostInstance): HostSshTarget | null {
  // IPv4 first: a v6-only target only works from a v6-capable client, and an
  // instance without either address is still being built.
  const address = instance.ipv4 ?? instance.ipv6;
  if (!address) return null;
  return {
    host: instance.defaultUser ? `${instance.defaultUser}@${address}` : address,
    name: instance.label,
    // `environment` is deliberately left unset. Vultr tags are free-form and
    // mean whatever the account owner decided, so inferring "Production" from
    // one would eventually put that label on the wrong machine. A user-saved
    // value wins in the merge anyway, so this is a blank the user can fill.
    instance: {
      providerId: VULTR_PROVIDER_ID,
      providerName: VULTR_MANIFEST.name,
      instanceId: instance.id,
      state: instance.state,
      rawState: instance.rawState,
      region: instance.region,
    },
  };
}

/** The read-safe half, over an already-authenticated `VultrManager`. */
export function createVultrHostProvider(manager: VultrManager): HostProvider {
  return {
    async account() {
      const account = await manager.account();
      return {
        // Vultr gives an account no id of its own; the email is what identifies
        // one, and is what its dashboard shows.
        id: account.email,
        username: account.email,
        name: account.name || null,
        email: account.email || null,
      };
    },
    async listInstances() {
      return (await manager.listInstances()).map(toHostInstance);
    },
    async getInstance(id) {
      return toHostInstance(await manager.getInstance(id));
    },
    toSshTarget: toVultrSshTarget,
  };
}

/**
 * The write-capable half. Kept a separate factory, resolved separately, and
 * never handed to an MCP tool — the same boundary `VultrActions` has.
 */
export function createVultrHostActions(actions: VultrActions): HostProviderActions {
  return {
    async run(action: string, instanceId: string): Promise<void> {
      switch (action) {
        case "start":
          await actions.start(instanceId);
          return;
        case "halt":
          await actions.halt(instanceId);
          return;
        case "reboot":
          await actions.reboot(instanceId);
          return;
        default:
          throw new Error(`Vultr does not support the action "${action}".`);
      }
    },
  };
}
